internal/unionstore/pipelined_memdb.go

```diff
@@ -0,0 +1,286 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unionstore
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/pingcap/errors"
+	tikverr "github.com/tikv/client-go/v2/error"
+	"github.com/tikv/client-go/v2/kv"
+	"github.com/tikv/client-go/v2/metrics"
+	"github.com/tikv/client-go/v2/util"
+)
+
+// PipelinedMemDB is a Store which contains
+//   - a mutable buffer for read and write
+//   - an immutable onflushing buffer for read
+//   - like MemDB, PipelinedMemDB also CANNOT be used concurrently
+type PipelinedMemDB struct {
+	// Like MemDB, this RWMutex only used to ensure memdbSnapGetter.Get will not race with
+	// concurrent memdb.Set, memdb.SetWithFlags, memdb.Delete and memdb.UpdateFlags.
+	sync.RWMutex
+	onFlushing        atomic.Bool
+	errCh             chan error
+	flushFunc         FlushFunc
+	bufferBatchGetter BufferBatchGetter
+	memBuffer         *MemDB
+	flushingMemBuffer *MemDB // the flushingMemBuffer is not wrapped by a mutex, because there is no data race in it.
+	len, size         int    // len and size records the total flushed and onflushing memdb.
+}
+
+const (
+	// MinFlushKeys is the minimum number of keys to trigger flush.
+	// small batch can lead to poor performance and resource waste in random write workload.
+	// 10K batch size is large enough to get good performance with random write workloads in tests.
+	MinFlushKeys = 10000
+	MinFlushSize = 16 * 1024 * 1024  // 16MB
+	MaxFlushSize = 128 * 1024 * 1024 // 128MB
+)
+
+type pipelinedMemDBSkipRemoteBuffer struct{}
+
+var pipelinedMemDBSkipRemoteBufferKey = pipelinedMemDBSkipRemoteBuffer{}
+
+// WithPipelinedMemDBSkipRemoteBuffer is used to skip reading remote buffer for saving RPC.
+func WithPipelinedMemDBSkipRemoteBuffer(ctx context.Context) context.Context {
+	return context.WithValue(ctx, pipelinedMemDBSkipRemoteBufferKey, struct{}{})
+}
+
+type FlushFunc func(*MemDB) error
+type BufferBatchGetter func(ctx context.Context, keys [][]byte) (map[string][]byte, error)
+
+func NewPipelinedMemDB(bufferBatchGetter BufferBatchGetter, flushFunc FlushFunc) *PipelinedMemDB {
+	return &PipelinedMemDB{
+		memBuffer:         newMemDB(),
+		errCh:             make(chan error, 1),
+		flushFunc:         flushFunc,
+		bufferBatchGetter: bufferBatchGetter,
+	}
+}
+
+// Dirty returns whether the pipelined buffer is mutated.
+func (p *PipelinedMemDB) Dirty() bool {
+	return p.memBuffer.Dirty() || p.len > 0
+}
+
+// GetMemDB returns the mutable memdb ONLY for writing.
+func (p *PipelinedMemDB) GetMemDB() *MemDB {
```

Comment:
Is this still needed if the `PipelinedMemDB` has implemented the same interface with `MemDB`?

Reply:
It's a sideway to access the inner memdb, and should be unreachable for pipelined memdb.

Because the usual `memBufferMutations` need to be built from `MemDB` objects, the original memdb interface is provided.

https://github.com/tikv/client-go/blob/b8855b2e80fa0dc76d9b4f504f9e9b22d5590cd6/txnkv/transaction/2pc.go#L549-L551